src/journal/compress.c

```diff
@@ -677,12 +865,102 @@ int decompress_stream_lz4(int in, int out, uint64_t max_bytes) {
 #endif
 }
 
+int decompress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeDCtxp) ZSTD_DCtx *s = ZSTD_createDStream();
+        size_t out_size = ZSTD_CStreamOutSize();
+        uint8_t out[out_size];
+        char *src = NULL;
+        struct stat st;
+        size_t ret;
+        uint64_t left = max_bytes;
+        int r;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        if (!s)
+                return -ENOMEM;
+
+        ret = ZSTD_initDStream(s);
+        if (ZSTD_isError(ret)) {
+                log_error("Failed to initialize ZSTD encoder: %s", ZSTD_getErrorName(ret));
+                return zstd_ret_to_errnoh(ret);
+        }
+
+        if (fstat(fdf, &st) < 0)
+                return log_debug_errno(errno, "fstat() failed: %m");
+
+        src = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE, fdf, 0);
+        if (src == MAP_FAILED)
+                return -errno;
+
+        for (;;) {
+                bool first = true;
+                ZSTD_inBuffer ZSTD_in = {src, st.st_size, 0};
+                ZSTD_outBuffer ZSTD_out = {out, out_size, 0};
+
+                if (first) {
+                        unsigned long long q;
+
+                        if (ZSTD_frameHeaderSize_prefix > ZSTD_in.size)
+                                return -EBADMSG;
+                        if (ZSTD_frameHeaderSize(ZSTD_in.src, ZSTD_in.size) > ZSTD_in.size)
+                                return -EBADMSG;
+                        q = ZSTD_findDecompressedSize(ZSTD_in.src, ZSTD_in.size);
```

Comment:
You should use `ZSTD_getFrameContentSize()` here, unless you are explicitly supporting multiple concatenated streams. It only needs to look at the first 15 bytes of the file at most, so you don't need to `mmap()`. The call to `ZSTD_frameHeaderSize()` also isn't necessary. With both of these calls gone, we no longer need `ZSTD_STATIC_LINKING_ONLY`.

Both of the compression functions produce exactly one compressed frame.